testing/test_metal_context.mm

```diff
@@ -14,46 +14,52 @@
 #include "third_party/skia/include/gpu/ganesh/mtl/GrMtlBackendContext.h"
 #include "third_party/skia/include/gpu/ganesh/mtl/GrMtlDirectContext.h"
 
+static_assert(__has_feature(objc_arc), "ARC must be enabled.");
+
 namespace flutter {
 
 TestMetalContext::TestMetalContext() {
-  auto device = fml::scoped_nsprotocol{MTLCreateSystemDefaultDevice()};
+  id<MTLDevice> device = MTLCreateSystemDefaultDevice();
   if (!device) {
     FML_LOG(ERROR) << "Could not acquire Metal device.";
     return;
   }
 
-  auto command_queue = fml::scoped_nsobject{[device.get() newCommandQueue]};
+  id<MTLCommandQueue> command_queue = [device newCommandQueue];
   if (!command_queue) {
     FML_LOG(ERROR) << "Could not create the default command queue.";
     return;
   }
 
-  [command_queue.get() setLabel:@"Flutter Test Queue"];
+  [command_queue setLabel:@"Flutter Test Queue"];
 
   GrMtlBackendContext backendContext = {};
   // Skia expect arguments to `MakeMetal` transfer ownership of the reference in for release later
   // when the GrDirectContext is collected.
-  backendContext.fDevice.reset([device.get() retain]);
-  backendContext.fQueue.reset([command_queue.get() retain]);
+  backendContext.fDevice.reset((__bridge_retained GrMTLHandle)device);
+  backendContext.fQueue.reset((__bridge_retained GrMTLHandle)command_queue);
   skia_context_ = GrDirectContexts::MakeMetal(backendContext);
   if (!skia_context_) {
     FML_LOG(ERROR) << "Could not create the GrDirectContext from the Metal Device "
                       "and command queue.";
   }
 
-  device_ = [device.get() retain];
-  command_queue_ = [command_queue.get() retain];
+  // Retain and transfer to non-ARC-managed pointers.
+  device_ = (__bridge_retained void*)device;
+  command_queue_ = (__bridge_retained void*)command_queue;
 }
 
 TestMetalContext::~TestMetalContext() {
   std::scoped_lock lock(textures_mutex_);
   textures_.clear();
   if (device_) {
-    [(__bridge id)device_ release];
+    // Release and transfer to ARC-managed pointer.
+    id<MTLDevice> device = (__bridge_transfer id<MTLDevice>)device_;
+    device = nil;
   }
   if (command_queue_) {
-    [(__bridge id)command_queue_ release];
+    id<MTLCommandQueue> command_queue = (__bridge_transfer id<MTLCommandQueue>)command_queue_;
```

Comment:
The clang docs say the operand "must have non-retainable pointer type", which isn't the case here; this seems like we are abusing the bridge casts in this code.

What's the high level goal here? Can we adjust things not to require manual reference count management and just extend the lifetime?

Reply:
Agreed, this is gross. My first attempt at this was to migrate these two fields to Objective-C pointers:
https://github.com/flutter/engine/blob/10e50bd42c42d144443806387713ca3ebdd584d9/testing/test_metal_context.h#L41-L42

However, this header is used from mostly plain C++ files and so, this causes us to have to migrate all of those to Objective-C++ as well. I didn't investigate how many files that would affect.

Reply:
Giving a go at the refactoring. It might not be that bad.

Reply:
I've done a test refactoring that special-cases Metal and it does require refactoring:
* EmbedderTestBackingStoreProducer to have backend-specific subclasses.
* EmbedderConfigBuilder, which instantiates the above.
* EmbedderTest, which holds onto EmbedderTestContextMetal, to have backend-specific subclasses.

Right now one test config can actually support multiple rendering backends. Practically speaking, I don't think's a big deal for Metal, since on macOS/iOS Metal and the software backend are the only ones we support, and I think we could still keep the software backends in the common superclass. If we wanted to refactor Vulkan/OpenGL into their own subclasses, we'd need to think about it further, since IIRC Android supports Vulkan, OpenGL, and Software.